core/server/worker/src/main/java/alluxio/worker/block/allocator/RoundRobinAllocator.java

```diff
@@ -79,31 +86,56 @@ private StorageDirView allocateBlock(long sessionId, long blockSize,
         int dirViewIndex = getNextAvailDirInTier(tierView, blockSize,
             BlockStoreLocation.ANY_MEDIUM);
         if (dirViewIndex >= 0) {
-          mTierAliasToLastDirMap.put(tierView.getTierViewAlias(), dirViewIndex);
-          return tierView.getDirView(dirViewIndex);
+          StorageDirView candidateDir = tierView.getDirView(dirViewIndex);
+          if (mReviewer.reviewAllocation(candidateDir)) {
+            mTierAliasToLastDirMap.put(tierView.getTierViewAlias(), dirViewIndex);
+            return candidateDir;
+          } else {
+            // Try the next dir
+            LOG.debug("Allocation rejected for anyTier: {}",
+                    candidateDir.toBlockStoreLocation());
+          }
         } else { // we didn't find one in this tier, go to next tier
           tierIndex++;
         }
       }
     } else if (location.equals(BlockStoreLocation.anyDirInTier(location.tierAlias()))) {
       StorageTierView tierView = mMetadataView.getTierView(location.tierAlias());
-      int dirViewIndex = getNextAvailDirInTier(tierView, blockSize, BlockStoreLocation.ANY_MEDIUM);
-      if (dirViewIndex >= 0) {
-        mTierAliasToLastDirMap.put(tierView.getTierViewAlias(), dirViewIndex);
-        return tierView.getDirView(dirViewIndex);
+      int offset = 0;
+      while (offset < tierView.getDirViews().size() && offset > -1) {
+        int dirViewIndex = getNextAvailDirInTier(tierView, blockSize, BlockStoreLocation.ANY_MEDIUM);
+        if (dirViewIndex >= 0) {
+          StorageDirView candidateDir = tierView.getDirView(dirViewIndex);
+          if (mReviewer.reviewAllocation(candidateDir)) {
+            mTierAliasToLastDirMap.put(tierView.getTierViewAlias(), dirViewIndex);
+            return tierView.getDirView(dirViewIndex);
+          } else {
+            // Reject the allocation
+            LOG.debug("Allocation to dirIndex {} rejected for anyDirInTier: {}", dirViewIndex,
+                    candidateDir.toBlockStoreLocation());
+          }
+        }
+        // Either the dirViewIndex is -1, or we rejected the allocation to a dirViewIndex > 0
+        offset = dirViewIndex;
       }
     } else if (location.equals(BlockStoreLocation.anyDirInTierWithMedium(location.mediumType()))) {
-      String medium = location.mediumType();
-      int tierIndex = 0; // always starting from the first tier
-      for (int i = 0; i < mMetadataView.getTierViews().size(); i++) {
```

Comment:
The previous code was wrong -- the quest is `anyDirInTierWithMedium` but here it is checking every tier. Maybe I should fix this in a separate PR.

Reply:
This is addressed separately in https://github.com/Alluxio/alluxio/pull/12509